### PR TITLE
[yaml] yaml_transform.py unit tests python - PR 3/3

### PR DESCRIPTION
**Pull request by bzablocki**
The last PR with Unit Tests. This PR covers YamlTransform class and most of the functions in yaml_transform.py that are not belonging to any class. This PR is meant to be reviewed after: 
1) https://github.com/apache/beam/pull/27086
2) https://github.com/apache/beam/pull/27338

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by bzablocki** [comment]
R: @robertwb

**Comment by bzablocki** [comment]
Hi @robertwb, could you also take a look at this PR?

File: sdks/python/apache_beam/yaml/yaml_transform_ut_test.py [hunk 1]
```diff
@@ -0,0 +1,1049 @@
...
+            elements: [1,2,3]
+          - type: PyMap
+            fn: 'lambda x: x*x'
+      '''
+    spec = yaml.load(spec, Loader=SafeLineLoader)
+    result = YamlTransform(spec)
+    self.assertIn('PyMap', result._providers)  # check for standard provider
+    self.assertEqual(result._spec['type'], "composite")  # preprocessed spec
+
+  def test_expand_pcollection(self):
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: removed

[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: removing

File: sdks/python/apache_beam/yaml/yaml_transform_ut_test.py [hunk 2]
```diff
@@ -0,0 +1,1049 @@
...
+
+  def test_ensure_transforms_have_types(self):
+    spec = '''
+      type: PyMap
+      fn: 'lambda x: x*x'
+      input: Flatten
+    '''
+    spec = yaml.load(spec, Loader=SafeLineLoader)
+    result = ensure_transforms_have_types(copy.deepcopy(spec))
+    self.assertCountEqual(result, spec)
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: Good point, I double-checked and these are dicts, so they can be compared with assertEqual.

File: sdks/python/apache_beam/yaml/yaml_transform_ut_test.py [hunk 3]
```diff
@@ -0,0 +1,1049 @@
...
+    self.assertIn("__line__", result_flatten)
+    self.assertIn("__uuid__", result_flatten)
+
+    self.assertEqual(result_pymap['input']['input'], result_flatten["__uuid__"])
+    self.assertEqual(result_pymap['type'], "PyMap")
+    self.assertEqual(result_pymap['fn'], 'lambda x: x*x')
+    self.assertIn("__line__", result_pymap)
+    self.assertIn("__uuid__", result_pymap)
+
+  def test_preprocess_flattened_inputs_explicit_flatten(self):
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: I wanted to test if the `preprocess_flattened_inputs()` behaves correctly if the spec has an explicit flatten transform defined. I can definitely simplify the checks.

File: sdks/python/apache_beam/yaml/yaml_transform_ut_test.py [hunk 4]
```diff
@@ -0,0 +1,1049 @@
...
+    result = preprocess_windowing(copy.deepcopy(spec))
+    self.assertCountEqual(spec, result)
+    self.assertEqual(spec['transforms'], result['transforms'])
+
+  def test_preprocess_windowing_other_type_with_no_inputs(self):
+    spec = '''
+      type: composite
+      transforms:
+        - type: CreateTimestamped
+          name: Create1
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: No reason, I'll rename to 'create'.

File: sdks/python/apache_beam/yaml/yaml_transform_ut_test.py [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+    spec = spec['transforms'][0]
+    result = preprocess_windowing(copy.deepcopy(spec))
+
+    # Get resulting WindowInto transform
+    result_window = self.get_transform_by_type(result, "WindowInto")
+
+    # Get resulting SumGlobally transform
+    result_create = self.get_transform_by_type(result, "CreateTimestamped")
+
+    self.assertEqual(result['type'], "composite")
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: Very good point, I'll do that.

File: sdks/python/apache_beam/yaml/yaml_transform_ut_test.py [hunk 6]
```diff
@@ -0,0 +1,1049 @@
...
+      output: SumGlobally
+    '''
+    spec = yaml.load(spec, Loader=SafeLineLoader)
+    spec = normalize_inputs_outputs(spec)
+    spec['transforms'] = [
+        normalize_inputs_outputs(t) for t in spec['transforms']
+    ]
+
+    spec_sum = spec['transforms'][1]
+    # Pass a copy of spec_sum, because preprocess_windowing modifies
```
[the remark this answers is not in the document]
Comment (bzablocki) [reply sdks/python/apache_beam/yaml/yaml_transform_ut_test.py]: Now that I simplified the tests to use a yaml representation as an expected output, this is no longer an issue

**Comment by bzablocki** [comment]
Hi @robertwb, for some reason I didn't notice the previous review comments you made. Sorry for that. I addressed them with the latest commits. Your previous comments are marked as outdated, because I changed name of the file with to `yaml_transform_unit_tests.py`.
Before another review round, let's merge #27338

**Comment by robertwb** [comment]
This is ready to be rebased on the prior PR that was just merged.

**Comment by bzablocki** [comment]
@robertwb Rebased to master

**Review by robertwb — APPROVED** [review]

**Pull request merged**